scipy/special/_convex_analysis.pxd

```diff
@@ -20,15 +22,26 @@ cdef inline double kl_div(double x, double y) noexcept nogil:
     else:
         return INFINITY
 
+@cython.cdivision(True)
 cdef inline double rel_entr(double x, double y) noexcept nogil:
+    cdef double ratio
     if isnan(x) or isnan(y):
         return NAN
-    elif x > 0 and y > 0:
-        return x * log(x / y)
-    elif x == 0 and y >= 0:
-        return 0
-    else:
+    if x <= 0 or y <= 0:
+        if x == 0 and y >= 0:
+            return 0
         return INFINITY
```

Comment:
As long as you are adding code to get the edge cases correct, you might as well also handle x = ±∞ and y = ±∞.  In all four possibilities, the function should return nan.

Reply:
My understanding is that handling out of domain cases by returning $\infty$ (or $-\infty$ for concave functions) was a deliberate choice in `_convex_analysis.pxd`, and this behavior is consistent across the functions defined in this file. See for instance the discussion in https://github.com/scipy/scipy/issues/3421. If we should change this behavior, I think it needs it's own discussion first, and if we proceed, the change should be done simultaneously for all functions in `_convex_analysis.pxd`. It shouldn't impact this PR.

One thing that may still be inconsistent is that when `x` and `y` are both $+\infty$, the output is `nan`, as @fancidev pointed out above. I think this can be addressed separately if needed.

Reply:
@steppi, I see, thanks.  I hadn't followed that discussion closely, so I was applying the usual rules that `inf/inf` and `inf - inf` return `nan`.